modal/_partial_function.py

```diff
@@ -922,3 +922,174 @@ def wrapper(
+sys.path.insert(0, '{working_dir}')
+
+# Import the function
+from {module_name} import {func_name}
+
+# Load arguments if provided
+args_file = sys.argv[1] if len(sys.argv) > 1 else None
+if args_file and os.path.exists(args_file):
+    with open(args_file, 'r') as f:
+        call_info = json.load(f)
+        args = call_info.get('args', [])
+        kwargs = call_info.get('kwargs', {{}})
+else:
+    args = []
+    kwargs = {{}}
+
+# Call the function
+{func_name}.local(*args, **kwargs)
+"""
+
+                # Get function details
+                module_name = original_func.__module__
+                func_name = original_func.__name__
+                working_dir = os.getcwd()
+
+                # Format the script
+                script_content = script_template.format(
+                    working_dir=working_dir, module_name=module_name, func_name=func_name
+                )
+
+                # Write the script to a temporary file
+                with tempfile.NamedTemporaryFile(mode="w", suffix="_torchrun_wrapper.py", delete=False) as f:
+                    f.write(script_content)
+                    script_path = f.name
+
+                # Save arguments to a temporary file (if any)
+                args_file = None
+                if args or kwargs:
+                    # Convert args/kwargs to JSON-serializable format
+                    # Note: This is a simplification and may not work for all argument types
+                    with tempfile.NamedTemporaryFile(mode="w", suffix="_args.json", delete=False) as f:
+                        json.dump({"args": list(args), "kwargs": kwargs}, f)
+                        args_file = f.name
+
+                try:
+                    # Build the torchrun command
+                    cmd = [
+                        sys.executable,
+                        "-m",
+                        "torch.distributed.run",
+                        f"--nnodes={n_nodes}",
+                        f"--nproc-per-node={nproc_per_node}",
+                        f"--node-rank={container_rank}",
+                        f"--master-addr={main_ip_addr}",
+                        f"--master-port={master_port}",
+                        script_path,
+                    ]
+
+                    if args_file:
+                        cmd.append(args_file)
+
+                    print(f"Container rank {container_rank}: Executing torchrun")
+                    print(f"Command: {' '.join(cmd)}")
+
+                    # Run torchrun
+                    result = subprocess.run(cmd, check=True)
+
+                finally:
+                    # Clean up temporary files
+                    os.unlink(script_path)
+                    if args_file:
+                        os.unlink(args_file)
+
+        # Wrap the function
+        if isinstance(obj, _PartialFunction):
+            # Replace the raw function with our wrapper
+            obj.raw_f = torchrun_wrapper
```

Comment:
This part is a little unconventional right? I wonder if it's going to wedge us on a specific implementation for our modal decorators.